// @teaxyz/lib/script/mod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SemVer = exports.Path = exports.DownloadError = exports.InstallationNotFoundError = exports.PackageNotFoundError = exports.PantryNotFoundError = exports.PantryParseError = exports.PantryError = exports.ResolveError = exports.RunError = exports.TeaError = exports.semver = exports.hacks = exports.porcelain = exports.plumbing = exports.hooks = exports.utils = exports.types = void 0;
require("./_dnt.polyfills.js");
require("./src/utils/misc.js");
const misc_js_1 = require("./src/utils/misc.js");
const host_js_1 = __importDefault(require("./src/utils/host.js"));
const semver_js_1 = __importStar(require("./src/utils/semver.js")), semver = semver_js_1;
exports.SemVer = semver_js_1.default;
exports.semver = semver;
const Path_js_1 = __importDefault(require("./src/utils/Path.js"));
exports.Path = Path_js_1.default;
exports.types = __importStar(require("./src/types.js"));
const pkg = __importStar(require("./src/utils/pkg.js"));
const error_js_1 = require("./src/utils/error.js");
Object.defineProperty(exports, "TeaError", { enumerable: true, get: function () { return error_js_1.TeaError; } });
const useConfig_js_1 = __importDefault(require("./src/hooks/useConfig.js"));
const useOffLicense_js_1 = __importDefault(require("./src/hooks/useOffLicense.js"));
const useCache_js_1 = __importDefault(require("./src/hooks/useCache.js"));
const useCellar_js_1 = __importStar(require("./src/hooks/useCellar.js"));
Object.defineProperty(exports, "InstallationNotFoundError", { enumerable: true, get: function () { return useCellar_js_1.InstallationNotFoundError; } });
const useMoustaches_js_1 = __importDefault(require("./src/hooks/useMoustaches.js"));
const usePantry_js_1 = __importStar(require("./src/hooks/usePantry.js"));
Object.defineProperty(exports, "PantryError", { enumerable: true, get: function () { return usePantry_js_1.PantryError; } });
Object.defineProperty(exports, "PantryParseError", { enumerable: true, get: function () { return usePantry_js_1.PantryParseError; } });
Object.defineProperty(exports, "PantryNotFoundError", { enumerable: true, get: function () { return usePantry_js_1.PantryNotFoundError; } });
Object.defineProperty(exports, "PackageNotFoundError", { enumerable: true, get: function () { return usePantry_js_1.PackageNotFoundError; } });
const useFetch_js_1 = __importDefault(require("./src/hooks/useFetch.js"));
const useDownload_js_1 = __importStar(require("./src/hooks/useDownload.js"));
Object.defineProperty(exports, "DownloadError", { enumerable: true, get: function () { return useDownload_js_1.DownloadError; } });
const useShellEnv_js_1 = __importDefault(require("./src/hooks/useShellEnv.js"));
const useInventory_js_1 = __importDefault(require("./src/hooks/useInventory.js"));
const hydrate_js_1 = __importDefault(require("./src/plumbing/hydrate.js"));
const which_js_1 = __importDefault(require("./src/plumbing/which.js"));
const link_js_1 = __importDefault(require("./src/plumbing/link.js"));
const install_js_1 = __importStar(require("./src/plumbing/install.js"));
const resolve_js_1 = __importStar(require("./src/plumbing/resolve.js"));
Object.defineProperty(exports, "ResolveError", { enumerable: true, get: function () { return resolve_js_1.ResolveError; } });
const hacks_js_1 = require("./src/utils/hacks.js");
const useSync_js_1 = __importDefault(require("./src/hooks/useSync.js"));
const run_js_1 = __importStar(require("./src/porcelain/run.js"));
Object.defineProperty(exports, "RunError", { enumerable: true, get: function () { return run_js_1.RunError; } });
const install_js_2 = __importDefault(require("./src/porcelain/install.js"));
const utils = {
    pkg, host: host_js_1.default, flatmap: misc_js_1.flatmap, validate: misc_js_1.validate, panic: error_js_1.panic, ConsoleLogger: install_js_1.ConsoleLogger
};
exports.utils = utils;
const hooks = {
    useCache: useCache_js_1.default,
    useCellar: useCellar_js_1.default,
    useConfig: useConfig_js_1.default,
    useDownload: useDownload_js_1.default,
    useFetch: useFetch_js_1.default,
    useInventory: useInventory_js_1.default,
    useMoustaches: useMoustaches_js_1.default,
    useOffLicense: useOffLicense_js_1.default,
    usePantry: usePantry_js_1.default,
    useShellEnv: useShellEnv_js_1.default,
    useSync: useSync_js_1.default,
};
exports.hooks = hooks;
const plumbing = {
    hydrate: hydrate_js_1.default,
    link: link_js_1.default,
    install: install_js_1.default,
    resolve: resolve_js_1.default,
    which: which_js_1.default
};
exports.plumbing = plumbing;
const porcelain = {
    install: install_js_2.default,
    run: run_js_1.default
};
exports.porcelain = porcelain;
const hacks = {
    validatePackageRequirement: hacks_js_1.validatePackageRequirement
};
exports.hacks = hacks;
__exportStar(require("./src/types.js"), exports);
