// @teaxyz/lib/script/src/hooks/useConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._internals = exports.ConfigDefault = void 0;
const dntShim = __importStar(require("../../_dnt.shims.js"));
const misc_js_1 = require("../utils/misc.js");
const host_js_1 = __importDefault(require("../utils/host.js"));
const Path_js_1 = __importDefault(require("../utils/Path.js"));
function ConfigDefault(env = dntShim.Deno.env.toObject()) {
    const prefix = (0, misc_js_1.flatmap)(env['TEA_PREFIX']?.trim(), x => new Path_js_1.default(x)) ?? Path_js_1.default.home().join('.tea');
    const pantries = env['TEA_PANTRY_PATH']?.split(":").compact(x => (0, misc_js_1.flatmap)(x.trim(), x => Path_js_1.default.abs(x) ?? Path_js_1.default.cwd().join(x))) ?? [];
    const cache = Path_js_1.default.abs(env['TEA_CACHE_DIR']) ?? prefix.join('tea.xyz/var/www');
    const isCI = boolize(env['CI']) ?? false;
    const UserAgent = (0, misc_js_1.flatmap)(getv(), v => `tea.lib/${v}`) ?? 'tea.lib';
    //TODO prefer 'xz' on Linux (as well) if supported
    const compression = !isCI && (0, host_js_1.default)().platform == 'darwin' ? 'xz' : 'gz';
    return {
        prefix,
        pantries,
        cache,
        UserAgent,
        options: {
            compression,
        },
        git: git(prefix, env.PATH)
    };
}
exports.ConfigDefault = ConfigDefault;
function getv() {
    if (typeof dntShim.Deno === 'undefined') {
        const url = new URL(require("url").pathToFileURL(__filename).href);
        const path = new Path_js_1.default(url.pathname).parent().parent().parent().join("package.json");
        const blob = dntShim.Deno.readFileSync(path.string);
        const txt = new TextDecoder().decode(blob);
        const { version } = JSON.parse(txt);
        return typeof version == 'string' ? version : undefined;
    }
}
const gt = dntShim.dntGlobalThis;
function useConfig(input) {
    // storing on globalThis so our config is shared across
    // potentially multiple versions of libtea being loaded in the same process
    if (!gt.xyz_tea_config || input) {
        gt.xyz_tea_config = input ?? ConfigDefault();
    }
    return { ...gt.xyz_tea_config }; // copy to prevent mutation
}
exports.default = useConfig;
function boolize(input) {
    switch (input?.trim()?.toLowerCase()) {
        case '0':
        case 'false':
        case 'no':
            return false;
        case '1':
        case 'true':
        case 'yes':
            return true;
    }
}
function reset() {
    return delete gt.xyz_tea_config;
}
function initialized() {
    return gt.xyz_tea_config !== undefined;
}
exports._internals = { reset, initialized, boolize };
/// we support a tea installed or system installed git, nothing else
/// eg. `git` could be a symlink in `PATH` to tea, which would cause a fork bomb
/// on darwin if xcode or xcode/clt is not installed this will fail to our http fallback above
//TODO be able to use our own git if installed
//NOTE however we don’t want to have to fully hydrate its env when libtea is initialized only when needed so…
function git(_prefix, PATH) {
    return usr();
    function usr() {
        // only return /usr/bin if in the PATH so user can explicitly override this
        const rv = PATH?.split(":")?.includes("/usr/bin") ? new Path_js_1.default("/usr") : undefined;
        /// don’t cause macOS to abort and then prompt the user to install the XcodeCLT
        //FIXME test! but this is hard to test without docker images or something!
        if ((0, host_js_1.default)().platform == 'darwin') {
            if (new Path_js_1.default("/Library/Developer/CommandLineTools/usr/bin/git").isExecutableFile())
                return rv;
            if (new Path_js_1.default("/Applications/Xcode.app").isDirectory())
                return rv;
            return; // don’t use `git`
        }
        return rv?.join("bin/git");
    }
}
